### PR TITLE
firebase auth integration

### LINKED ISSUE #1332: Firebase Authentication

### PR DESCRIPTION
**Pull request by adamsachs**
# Purpose
Firebase Auth (https://firebase.google.com/products/auth) connector - access and update. Leverages the [Firebase Admin SDK](https://firebase.google.com/docs/auth/admin/manage-users) for integration.

# Changes
- Firebase connector config, dataset, test artifacts, etc. -- basic SaaS connector artifacts
- Request overrides which define the integration functionality using the Firebase Admin SDK
- Introduces the Firebase Admin SDK as a project dependency

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [ ] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [x] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #1332

### CODE REVIEW

File: data/saas/saas_connector_registry.toml
Comment (adamsachs) [inline data/saas/saas_connector_registry.toml:22]: not sure if we want to put an image in here or just use the default...

File: data/saas/config/firebase_auth_config.yml
Comment (adamsachs) [inline data/saas/config/firebase_auth_config.yml:42]: decided to configure `read` and `update` endpoints for now; we could easily implement `delete` (and i have in fact implemented the override function), but chose to keep it unconfigured (i.e. disabled) for now to avoid any confusion. open to discussing and/or adjusting.

File: requirements.txt
Comment (adamsachs) [inline requirements.txt:15]: not sure if anything else is needed to bring in this dependency, but things seem to work with just this update locally

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [inline src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:101]: as noted above, this isn't actually being leveraged currently. but i figured i'd keep the function implementation here just for reference, in case we do want to go down this route at some point. happy to include this in some other manner, but didn't want this work to get lost.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [inline src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:48]: i was having a bit of trouble testing this functionality because i don't think we're able to programmatically create provider data on a given user, so we can't seed the test data for this. i'll look to see if i can manually verify the functionality moving forward.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [inline src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:87]: `provider_data` cannot be updated programmatically via the API, so we don't try to update it here.

**Comment by adamsachs** [comment]
this may need a few more iterations, but i wanted to get an initial cut in front of some of your eyes. @seanpreston i put you on this just to make sure everything looks OK with bringing in the new `firebase_admin` dependency. 

@galvana / @earmenda  i've added test credentials to 1pw, could you help me add them to vault so that we can get the integration tests running on github?

**Comment by adamsachs** [comment]
the firebase test task is passing now, i'm getting an error with sendgrid that is almost certainly not related to these changes

**Comment by adamsachs** [comment]
now getting an error with braze in the saas tests, which i also doubt is related to the PR

File: data/saas/config/firebase_auth_config.yml
Comment (earmenda) [reply data/saas/config/firebase_auth_config.yml:42]: commented on this below

**Review by earmenda — COMMENTED** [review]
changes look good! just a few comments

File: data/saas/saas_connector_registry.toml
Comment (earmenda) [reply data/saas/saas_connector_registry.toml:22]: don't think this is important atm

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:101]: yup, that's how i felt about it too. maybe it's worth me throwing in a comment here just to clarify it moving forward.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:48]: interesting, thanks for digging that up! i'm with you on keeping the tests self-contained (and trying to run a CLI command within a pytest harness doesn't sound very appealing), but evaluating the provider data functionality may end up being important...so i'm kinda split here. i'll try to do some more digging on my end, thanks for the pointer.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:87]: that's not a bad point. perhaps it may depend on the use case - maybe we need some input from our first use case, and generally we should look to enable both pieces of functionality (as i've done by defining `update` and `delete` override functions)

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:58]: good catch! the problems that come with not having test coverage here...🤦

File: data/saas/dataset/firebase_auth_dataset.yml
Comment (adamsachs) [reply data/saas/dataset/firebase_auth_dataset.yml:1]: yeah, i'm with you on the `photo_url` - funny enough, i'd thought of including that in the dataset but then couldn't come up with a suitable fideslang data category to assign it, so then i just excluded it. but that's not a very good reason, so i think i'll include it again. any thoughts on what would be a good data category in the default taxonomy to assign it? (no worries if you don't have something off the top of your head)

the custom claims piece looks interesting - thanks for digging into that - but i'm just not sure how we'd handle that generically, since it's effectively schema-less. i _could_ see it containing relevant PII in certain use cases, so i don't feel great about just ignoring it

if we did decide to support this, what i'm thinking is we could allow users to configure their dataset according to what their `custom_claims` metadata looks like. then i could probably have something generic in the function to dynamically support whatever fields the dataset defines in the sub-object.

that's not a trivial amount of work, though, and we don't have a great paradigm in place for allowing users to overwrite the default datasets we're providing them, which would be necessary for this to work. so i'm a bit unsure here.

File: data/saas/config/firebase_auth_config.yml
Comment (galvana) [inline data/saas/config/firebase_auth_config.yml:33]: Should this be `query_params`?

File: data/saas/dataset/firebase_auth_dataset.yml
Comment (galvana) [reply data/saas/dataset/firebase_auth_dataset.yml:1]: I usually go with a more generic data category if a specific one isn't available, in this case I would lean towards just annotating `photo_url` as `user` data.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (galvana) [inline src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:131]: I remember you saying that we would get an error if we tried to initialize an "app" if one already existed. Is that why we're doing the try block? Not sure if I understand the issue but would Python's `global` keyword help here? We use it in `cache.py` to always return a single instance of the Redis cache.

File: data/saas/config/firebase_auth_config.yml
Comment (adamsachs) [reply data/saas/config/firebase_auth_config.yml:33]: good catch! i didn't mean to commit this, it was from an earlier iteration.

you know this actually isn't used at all - none of the `client_config` is. i don't think i can remove that wholesale, but i can remove the `authentication` since that's optional. i'll put in dummy values for the `protocol` and `host` to make it clearer that it's not being used.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:131]: actually, this is sort of the opposite of what i'd described. after mentioning the piece about initializing an app if it already exists, i found this method in the SDK that allows us to effectively retrieve an app by name/id, if it already exists; if it doesn't exist, it throws an error. and so we catch that here and initialize if it doesn't exist yet.

the `global` keyword may be a bit trickier here, because we don't want to assume there's only a single firebase `app` per fidesops runtime - that would prevent a single fidesops instance from integrating with multiple firebase projects, which is a limitation we should avoid if possible imo. perhaps we could maintain a global map of firebase `app`s tied to particular saas config instances, or something of the sort -- but that's effectively what we're getting for free here with the admin SDK!

i don't love that we need to try/catch, but other than that, i actually find this a pretty clean implementation. let me know what you think.

File: data/saas/dataset/firebase_auth_dataset.yml
Comment (adamsachs) [reply data/saas/dataset/firebase_auth_dataset.yml:1]: @earmenda and i discussed offline that we won't support `custom_claims` in our first iteration, but if it turns out to be an important client need, we can look to implement it.

we'd probably look to support it by either writing some code into our function to dynamically traverse/enumerate the subfields within the `custom_claims` field as declared in the dataset. we could also likely just serialize and dump out the `custom_claims` subobject as a JSON blob in an access request. we can look to explore those if/when the need arises.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:52]: noting that i found an SDK [function](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth?authuser=2#import_users)/[class](https://firebase.google.com/docs/reference/admin/python/firebase_admin.auth?authuser=2#importuserrecord) that allows us to import users with provider data defined. i'm updating test fixtures to create our sample users with provider data and include those pieces of data in the tests.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:91]: we're going to try to include some test coverage for both update and delete functionality, and keep both override functions in the codebase. that way it will essentially be a config switch (i.e. updating the saas config) whether to use update or delete in any given use case. we can adjust the default if needed, once we get some feedback from the real-world use cases.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (adamsachs) [inline src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:50]: this is going to add `None` values in for fields that are not populated in the response record dictionary/object. is that OK? i _think_ the behavior in "normal" saas requests is just to not include the field in the response record, in which case i think i'll need to handle this slightly differently...

@galvana @earmenda

**Comment by adamsachs** [comment]
@earmenda i think i've addressed all that we talked about. still outstanding questions on whether we'll need to handle the `client_claims` data and whether we want to default to `update` or `delete`, but per our discussion, we're not going to hold up merging on those and we can come back to them if/when needed.

i do still have one minor question (https://github.com/ethyca/fidesops/pull/1399/files#r985340851) that i'd like to clear up before merging. @galvana maybe you can confirm the general behavior there? or @earmenda if you know off the top of your head.

File: src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py
Comment (earmenda) [reply src/fidesops/ops/service/saas_request/override_implementations/firebase_auth_request_overrides.py:50]: i'd say use `hasattr` to avoid this

**Review by earmenda — APPROVED** [review]
lgtm!

**Review by galvana — APPROVED** [review]
A lot of interesting use cases with this connector and an overall good example of request overrides, approved!

**Pull request merged**
